In [2]:
%matplotlib inline

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd


from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

# Notes about softmax regression

Softmax regression aslo known as Mutinomial Logistic,Maximum entropy Clasifier, or just Multi-class logistic regression is a generelazation of logistic regression that we can use for multi-class classification(under the assumption that the classes are mutualy exclusive).<br>
<br>
In softmax regression we replace sigmoid function by so-called softmax function $\phi_{softmax}(.).$ <br> <br>
$$ P(y=j|z^{(i)}) = \phi_{softmax}(z^{(i)}) = \frac{e^{z^{(i)}}} {\sum_i^k e^{z_k^{(i)}}}$$ 
<br> <br>
Where we difine the net input as $$ z = w_1 x_1 + ...+w_nx_n +b = \sum_l^mw_lx_l +b = W^T +b$$
<br>
(W is the weight vector, X is is the feature of 1 training example, and b is bias unit )
<br>

Now,softmax function computes the probability that this training example $x^{(i)}$ belongs to class j given the weight and net input $z^{(i)}$. <br>
So we compute the probability $p(y=j | x^{(i)}) =;W_j$)$ for each clas label in j=1,...,k..Note the normalization term in the denominator which causes these class probabilities to sum up one.

To ilustrate the concept of softmax, let us walk through a concrete example. let's assume we have a training set consisting of 4 samples from 3 different classes (0, 1 and 2)
* $x_0 \rightarrow class \; 0 $ 
* $x_1 \rightarrow class \; 1 $ 
* $x_2 \rightarrow class \; 2 $ 
* $x_3 \rightarrow class \; 2 $ 

In [4]:
 y = np.array([0, 1, 2, 2])

First, we want to encode the class label into a format that we can more easily work with; we apply one-hot encoding :
<br>

In [21]:
y_enc = (np.arange(np.max(y)+1) == y [: , None]).astype(float)
print('one hot encoding')
y_enc

one hot encoding


array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.],
       [0., 0., 1.]])

In [20]:
 (np.array([1,2])==np.array([3,2])).astype(float)   ###### see


array([0., 1.])

The sample that belongs to class 0 (the first row) has a 1 in the first cell, a sampe that belong to class 2 has 1 in the second row, and so forth.
<br>
Next, let us define the feature of our 4 trainning samples.Here, we assume that our dataset consist of 2 features;thus, we create (4,2) dimensinal matrix of our samples and features. Similarly, we create a (2,3) dimensinal weigh matrix (one row per feature and one column per each class) 

In [22]:
X = np.array([[0.1, 0.5],
              [1.1, 2.3],
              [-1.1, -2.3],
              [-1.5, -2.5]])

W = np.array([[0.1, 0.2, 0.3],
              [0.1, 0.2, 0.3]])

bias = np.array([0.01, 0.1, 0.1])

print('Inputs X:\n', X)
print('\nWeights W:\n', W)
print('\nbias:\n', bias)

Inputs X:
 [[ 0.1  0.5]
 [ 1.1  2.3]
 [-1.1 -2.3]
 [-1.5 -2.5]]

Weights W:
 [[0.1 0.2 0.3]
 [0.1 0.2 0.3]]

bias:
 [0.01 0.1  0.1 ]


To compute the net input, we mutiply feature vector with weight vector with the $2\times 3$ (n_feature,n_classes) weight matrix W, wich yields a $4\times 3 $ matrix (n_samples,n_classes) to wich we then add the bias unit.
$$ Z = XW +b $$

In [23]:
X = np.array([[0.1, 0.5],
              [1.1, 2.3],
              [-1.1, -2.3],
              [-1.5, -2.5]])

W = np.array([[0.1, 0.2, 0.3],
              [0.1, 0.2, 0.3]])

bias = np.array([0.01, 0.1, 0.1])

print('Inputs X:\n', X)
print('\nWeights W:\n', W)
print('\nbias:\n', bias)

Inputs X:
 [[ 0.1  0.5]
 [ 1.1  2.3]
 [-1.1 -2.3]
 [-1.5 -2.5]]

Weights W:
 [[0.1 0.2 0.3]
 [0.1 0.2 0.3]]

bias:
 [0.01 0.1  0.1 ]


In [24]:
def net_input(X, W, b):
    return (X.dot(W) + b)

net_in = net_input(X, W, bias)
print('net input:\n', net_in)

net input:
 [[ 0.07  0.22  0.28]
 [ 0.35  0.78  1.12]
 [-0.33 -0.58 -0.92]
 [-0.39 -0.7  -1.1 ]]


Now, it's time to compute the softmax activation that we discussed earlier : <br> <br>
 $$ P(\; y=j\;|\;z^{(i)}) = \phi_{softmax}(z^{(i)}) = \frac{e^{z^{(i)}}} {\sum_i^k e^{z_k^{(i)}}} $$

In [26]:
def softmax(z):
    return (np.exp(z.T) / np.sum(np.exp(z), axis=1)).T

smax = softmax(net_in)
print('softmax:\n', smax)
# def sofmax(x,Z,intercept):
#     exp_K = np.exp(x.dot(Z.T)+intercept)
#     sums=np.sum(exp_K, axis=1) # array contains sum of exp if every row
#     result = (exp_K.T/sums).T  #return for every element e^{k_ij}/(e^{k_i1}+e^{k_i2}+e^{k_i3})
#     return result
    

softmax:
 [[0.29450637 0.34216758 0.36332605]
 [0.21290077 0.32728332 0.45981591]
 [0.42860913 0.33380113 0.23758974]
 [0.44941979 0.32962558 0.22095463]]


As we can see, the values for each sample (row) nicely sum up 1 now.E.g. , we can say the first tsample
$[0.29450637 0.34216758 0.36332605]$ has a $29.45 % $ probability to belong to class 0.

In [28]:
def to_classlabel(z):
    return z.argmax(axis=1) #return the index of max number of every row  in list

print('predicted class labels: ', to_classlabel(smax))

predicted class labels:  [2 2 0 0]


In [29]:
def to_classlabel(z):
    return z.argmax(axis=1)

print('predicted class labels: ', to_classlabel(smax))

predicted class labels:  [2 2 0 0]


As we can see, our prediction are terribly wrong the correct class labels are $[0,1,2,2]$.Now, in order to train out logistic model(e.g via the optimization algorithm such as gradient descent), we need to cost function &J& that we want to minimize : <br> <br>
$$J(W:b) = \frac{1}{n}\sum_i^nH(T_i,O_i), $$
Here the T stands for 'target' i.e the true class labels. <br>

$$ \nabla w_j J(W;b) = \frac{1}{n}\sum_i^ a$$
http://rasbt.github.io/mlxtend/user_guide/classifier/SoftmaxRegression/

in order to learn our sofmax model -- determening the weigth coefficient --via gradient descent, we then need to compute the derivative <br> <br>
   $$\nabla w_j J(W;b) $$
   
   
   
   final result look like this :
   <br> <br>
   
   $$\nabla w_j J(W;b) = frac{1}{n} \sum_i=0^n[x(i)(O_i-T_i)] $$